base-orphans doesn't build: Saw: `Data.Orphans.Prelude' Expected: `Prelude'

WORKSPACE:
```python
new_http_archive(
    name = "haskell_base_orphans",
    build_file = "third_party/haskell/BUILD.base-orphans",
    sha256 = "c7282aa7516652e6e4a78ccdfb654a99c9da683875748ad5898a3f200be7ad0e",
    strip_prefix = "base-orphans-0.6",
    urls = ["https://hackage.haskell.org/package/base-orphans-0.6/base-orphans-0.6.tar.gz"],
)
```

third_party/haskell/BUILD.base-orphans:
```python
load("@io_tweag_rules_haskell//haskell:haskell.bzl", "haskell_library")

haskell_library(
    name = "base-orphans",
    srcs = glob(["src/**/*.hs*"]),
    prebuilt_dependencies = ["base"],
    src_strip_prefix = "src",
    visibility = ["//visibility:public"],
)
```

Error:
```
external/haskell_base_orphans/src/Data/Orphans/Prelude.hs:17:8: error:
    File name does not match module name:
    Saw: `Data.Orphans.Prelude'
    Expected: `Prelude'
Target @haskell_base_orphans//:base-orphans failed to build
```

--- a/haskell/actions.bzl
+++ b/haskell/actions.bzl
@@ -4,6 +4,7 @@ load(":path_utils.bzl",
      "declare_compiled",
      "target_unique_name",
      "module_name",
+     "import_hierarchy_root",
 )
 
 load(":set.bzl", "set")
@@ -505,10 +506,9 @@ def compilation_defaults(ctx):
 
   _add_mode_options(ctx, args)
 
-  for idir in set.to_list(set.from_list([f.dirname for f in sources])):
-    items = ["-i{0}".format(idir)]
-    args.add(items)
-    haddock_args.add(items, before_each="--optghc")
+  ih_root_arg = ["-i{0}".format(import_hierarchy_root(ctx))]
+  args.add(ih_root_arg)
+  haddock_args.add(ih_root_arg, before_each="--optghc")
 
   dep_info = gather_dependency_information(ctx)
 

--- a/haskell/path_utils.bzl
+++ b/haskell/path_utils.bzl
@@ -84,6 +84,25 @@ def declare_compiled(ctx, src, ext, directory=None):
   fp_with_dir = fp if directory == None else paths.join(directory.basename, fp)
   return ctx.actions.declare_file(fp_with_dir)
 
+def import_hierarchy_root(ctx):
+  """Return relative path to root of module hierarchy.
+
+  Args:
+    ctx: Rule context.
+
+  Returns:
+    string: Relative path to root of module hierarchy.
+  """
+  return paths.join(
+    ctx.label.workspace_root,
+    ctx.label.package,
+    # Since the src_strip_prefix attribute is always present in rule
+    # attributes, if it's not there, the function is called from aspect
+    # implementation and so we can access ctx.rule.attr.src_strip_prefix.
+    ctx.attr.src_strip_prefix if hasattr(ctx.attr, "src_strip_prefix")
+                              else ctx.rule.attr.src_strip_prefix
+  )
+
 def _rel_path_to_module(ctx, f):
   """Make given file name relative to the directory where the module hierarchy
   starts.
@@ -99,18 +118,7 @@ def _rel_path_to_module(ctx, f):
   Returns:
     string: Relative path to module file.
   """
-  return paths.relativize(
-    f.path,
-    paths.join(
-      ctx.label.workspace_root,
-      ctx.label.package,
-      # Since the src_strip_prefix attribute is always present in rule
-      # attributes, if it's not there, the function is called from aspect
-      # implementation and so we can access ctx.rule.attr.src_strip_prefix.
-      ctx.attr.src_strip_prefix if hasattr(ctx.attr, "src_strip_prefix")
-                                else ctx.rule.attr.src_strip_prefix
-    )
-  )
+  return paths.relativize(f.path, import_hierarchy_root(ctx))
 
 def _drop_extension(f):
   """Drop extension for a given file name.

